src/rascal/structure_managers/structure_manager.hh

```diff
@@ -400,88 +447,162 @@ namespace rascal {
-      auto property = this->get_property(name);
-      this->template validate_property_t<UserProperty_t>(property);
-      return std::static_pointer_cast<UserProperty_t>(property);
-    }
-
-    void register_property(std::shared_ptr<PropertyBase> property,
-                           const std::string & name) {
-      this->properties[name] = property;
-    }
-
-    /**
-     * Get a atom property of a given name, possibly excluding ghosts. Create it
-     * if it does not exist.
-     *
-     * @tparam UserProperty_t full type of the property to return
+     * Returns a typed property of the given name.
      *
-     * @param name name of the property to get
+     * @tparam UserProperty_t full type of the property to return.
      *
+     * @param name the name of the property to get.
+     * @param validate_property property is validated if this parameter is true,
+     * see validate_property_t
+     * @param force_creation if the property does not exist in the manager stack
+     * the property is created and returned. The validation step is skipped in
+     * this case.
      * @param exclude_ghosts change property sizing behavior when Order == 1
      * and when the property does not already exist.
+
      *
-     * @throw runtime_error if UserProperty_t is not compatible with property
-     * of the given name
+     * @throw runtime_error If validate_property is true and UserProperty_t is
+     * not compatible with the property with the given name.
+     * @throw runtime_error If force_creation is false and property has not been
+     * found in manager stack.
      */
     template <typename UserProperty_t>
-    std::shared_ptr<UserProperty_t>
-    get_property_ptr(const std::string & name,
-                     const bool & exclude_ghosts = false,
-                     const std::string & metadata = "no metadata") {
-      if (this->has_property(name)) {
-        auto property{this->get_property(name)};
-        UserProperty_t::check_compatibility(*property);
-        return std::static_pointer_cast<UserProperty_t>(property);
-      } else {
+    std::shared_ptr<UserProperty_t> get_property(
+        const std::string & name, const bool validate_property = true,
+        const bool force_creation = false, const bool exclude_ghosts = false,
+        const std::string & metadata = "no metadata") {
+      if (this->has_stack_property(name)) {
+        return this->template forward_get_property_request<UserProperty_t>(
+            name, validate_property);
+      }
+      if (force_creation) {
         auto property{std::make_shared<UserProperty_t>(
             this->implementation(), metadata, exclude_ghosts)};
-        this->register_property(property, name);
+        this->properties[name] = property;
         return property;
       }
+      std::stringstream error{};
+      error << "No property of name '" << name << "' has been registered";
+      throw std::runtime_error(error.str());
     }
```

Comment:
the logic seems a bit funky here. Also I would set `const bool force_creation = true` as default since it would be the case most of the times. 
How about:
```
if (this->has_stack_property(name)) {
...
} else if (not this->has_stack_property(name) and force_creation) { 
...
} else {
      std::stringstream error{};
      error << "No property of name '" << name << "' has been registered";
      throw std::runtime_error(error.str());
}
```

Reply:
I don't see the benefit in readability.